tool/site/client-react/src/components/longRunningTask.tsx

```diff
@@ -0,0 +1,75 @@
+import * as React from "react"
+import {JSXElementConstructor, useContext, useEffect, useState} from "react"
+import {Button, ErrorMessage, Label} from "nhsuk-react-components"
+import ButtonList from "./buttonList"
+import {AppContext} from "../index"
+
+interface LongRunningTaskProps<T> {
+  task: () => Promise<T>
+  loadingMessage: string
+  children: JSXElementConstructor<T>
+  back?: string | (() => void)
+}
+
+const LongRunningTask = <T extends unknown>({
+  task,
+  loadingMessage,
+  children,
+  back
+}: LongRunningTaskProps<T>): React.ReactElement => {
+  const {baseUrl} = useContext(AppContext)
+  const [loading, setLoading] = useState<boolean>(true)
+  const [errorMessage, setErrorMessage] = useState<string>()
+  const [result, setResult] = useState<T>()
+
+  if (!back) {
+    back = baseUrl
+  }
+
+  useEffect(() => {
+    async function performTask() {
+      setLoading(true)
+      try {
+        const loadResult = await task()
+        setResult(loadResult)
+      } catch (e) {
+        console.log(e)
+        setErrorMessage(e.message)
+      } finally {
+        setLoading(false)
+      }
+    }
+
+    if (!result) {
+      performTask()
+    }
+  }, [result, task])
+
+  if (errorMessage) {
+    return (
+      <>
+        <Label isPageHeading>Error</Label>
+        <ErrorMessage>{errorMessage}</ErrorMessage>
+        <ButtonList>
+          {typeof back === "string"
```

Comment:
Why is back either a string or a function?

Reply:
Sometimes going back means navigating to a different page. Other times going back means setting some state within the parent component. So this handles either case